R/db.R

```diff
@@ -264,12 +457,20 @@ build_insert_query <- function(tblname, df) {
 #' @importFrom DBI dbGetQuery
 get_new_key <- function(dbconn, tbl_name, ids, key) {
   if (length(ids) == 0) return(integer(0))
-  if (!DBI::dbExistsTable(dbconn, tbl_name)) return(ids)
+  shards <- get_shards_for_table(dbconn, tbl_name)
+  ## If there are no existing shards - then nothing is cached yet
+  if(NROW(shards) == 0) {
+    return(ids)
+  } else {
+    shards <- shards[[1]]
+  }
+  if (!DBI::dbExistsTable(dbconn, shards[1])) return(ids)
   id_column_name <- get_hashed_names(key)
+  ## We can check only the first shard because all shards have the same keys
```

Comment:
Yeah, since it's transactional I think this is the correct way. The set of present keys should be consistent amongst shards by design.